security/pkg/nodeagent/secretfetcher/secretfetcher.go

```diff
@@ -144,17 +145,69 @@ func (sf *SecretFetcher) Init(core corev1.CoreV1Interface) { // nolint:interface
 		})
 }
 
-func extractCertAndKey(scrt *v1.Secret) (cert, key, root []byte) {
+func extractCertAndKey(scrt *v1.Secret) (cert, key []byte) {
 	if len(scrt.Data[genericScrtCert]) > 0 {
 		cert = scrt.Data[genericScrtCert]
 		key = scrt.Data[genericScrtKey]
-		root = scrt.Data[genericScrtCaCert]
 	} else {
 		cert = scrt.Data[tlsScrtCert]
 		key = scrt.Data[tlsScrtKey]
-		root = []byte{}
 	}
-	return cert, key, root
+	return cert, key
+}
+
+// extractCACert extracts the client CA certificate from either the Opaque
+// Secret, or, if legacyOnly=false, from k8s's "tls.crt".
+func extractCACert(scrt *v1.Secret, legacyOnly bool) (cacert []byte) {
```

Comment:
> it is still useful for use cases where cert-manager is not deployed

Arguably nothing stops an operator from deploying a server certificate and a client certificate in two secrets. The decision to name it as legacy is for the sake of separation of concerns and to promote the further deprecation: the existing code demonstrates UB when a secret has a suffix of "-cacert" and the new code doesn't solve the UB problem. I'd think it's for the better to have a non-conflicting secret resolution that doesn't have any naming conflicts – which would require deprecating the original code at some point (or banning the secrets with a suffix of "-cacert" – but that seems counterproductive).

What do you think?

Reply:
I get you point basically, except that I don't know what UB stands for. I think adding a brief comment to explain the context above the function would be nice, so that other readers are clear what cases extractCACert() is dealing with.

Reply:
Undefined Behavior :)

When nodeagent faces a secret named `test` and then a secret named `test-cacert` it can break down in a ways that no one anticipated and that's not clearly documented anywhere.

Reply:
Thanks for the explanation. 
Perhaps we can add some checks in a follow up PR like this.
(1) If a compound secret "example" is created, and operator wants to add "example-cacert", node agent generates an error message and skip adding "example-cacert". Operator must delete compound "example" first, and create "example" and "example-cacert".
(2) if a secret "example-cacert" is created, and operator wants to add a compound "example", node agent generates an error message and skip adding "example". Operator must delete "example-cacert" first, and then create compound "example".

Reply:
That's a good idea. I think it will require the nodeagent to be more aware of the Secret to SecretItem mapping, though, but it might be a good thing.